ARROW-835: [Format][C++][Java] Create a new Duration type

emkornfield opened the pull request:
- Create a new DurationInterval type in format.
- Implement all interval types in C++ and Java including integration test.

Once this is checked in, I think https://issues.apache.org/jira/browse/ARROW-352 can be resolved as a  won't fix?

emkornfield commented:
@wesm could you look at the format changes to see if you are ok with them?
@pravindra Can you look at the impact on the java side (I added all the necessary java classes I think) to confirm this is acceptable for the Java?

emkornfield commented on format/Schema.fbs:
```diff
@@ -183,10 +186,25 @@ table Timestamp {
+// An "calendar" interval which models types that don't necessarily
+// have a precise description.
+// YEAR_MONTH - Indicates the number of elapsed whole months.
+// DAY_TIMME - Indicates an 8-byte wide struct containing 2 bit integers
```
Sorry missed the e-mail on this.  Will fix in a second.  My intent was not to change the java representation which appears to store Days and Tiem separately.

emkornfield replied:
I pushed a new commit, please take a look.

emkornfield replied:
Sorry just addressed this comment more fully.

emkornfield commented on java/vector/src/main/java/org/apache/arrow/vector/IntervalEpochMilliVector.java:
```diff
@@ -0,0 +1,405 @@
+
+/**
+ * IntervalEpochMilliVector implements a fixed width vector (8 bytes) of
+ * a second  granularity values which could be null.
```
Done.  Tried to reword a little bit to make more sense.

emkornfield commented:
Failure was NodeJs

pravindra commented:
since this is a format change (backward compatible though), I'll wait for @wesm to approve too before merging.

emkornfield commented:
@pravindra thanks for the review

emkornfield commented on format/Schema.fbs:
```diff
@@ -183,10 +183,27 @@ table Timestamp {
+/// millisecond, but can be any of the other supported TimeUnit values as
+/// with Timestamp and Time types.  This type is always represented as
+/// an 8-byte integer.
+table DurationInterval {
```
Looking over the old e-mail threads it might be that TimeDelta or TimestampDelta is a better name for this type?

wesm replied:
Are you happy with "DurationInterval"? It could also be called simply "Duration" or "TimeDuration"

emkornfield commented:
@wesm based on comments on the other JIRAs CLs it seems like it would be desirable to add an implementation for C++, I can add that to this CL but would you mind taking a look to see if you are ok with naming/modeling (I would prefer not to have to refactor just due to naming issues if possible).

emkornfield commented:
I believe the Python on OS/X error is spurious (it was green on previous build).

wesm commented:
Will work on reviewing the C++ side of this

wesm commented on cpp/src/arrow/array.h:
```diff
@@ -650,6 +650,36 @@ class ARROW_EXPORT FixedSizeBinaryArray : public PrimitiveArray {
 
+/// DayTimeArray
+/// ---------------------
+/// \brief Array of Day and Millisecond values.
```
FWIW, I'm a bit disappointed that we aren't deprecating this type =/

emkornfield replied:
I didn't want to get dragged into what it means deprecate this, should be easy enough to do if we want it.  As noted below I tried to say it was optional which I seem to recall seeing this was ok on a previous ML thread.

wesm commented on cpp/src/arrow/array/builder_time.h:
```diff
@@ -0,0 +1,77 @@
+using Time32Builder = NumericBuilder<Time32Type>;
+using Time64Builder = NumericBuilder<Time64Type>;
+using Date32Builder = NumericBuilder<Date32Type>;
+using Date64Builder = NumericBuilder<Date64Type>;
```
Is this redundant with `type_fwd.h`?

emkornfield replied:
Yes, it appears so, I'll push a commit removing them.  It looks like there might be more redundancy in builder_primitive.h

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -111,6 +112,18 @@ Status MakeBuilder(MemoryPool* pool, const std::shared_ptr<DataType>& type,
       return VisitTypeInline(*dict_type.dictionary()->type(), &visitor);
     }
+    case Type::INTERVAL: {
+      const auto& interval_type = dynamic_cast<const IntervalType&>(*type);
```
I think `static_cast` (or `checked_cast`) is sufficient here?

wesm commented on cpp/src/arrow/type.h:
```diff
@@ -890,6 +950,15 @@ std::shared_ptr<DataType> list(const std::shared_ptr<Field>& value_type);
+std::shared_ptr<DataType> ARROW_EXPORT day_time_interval();
+
+/// \brief Return an MonthIntervalType instance
+std::shared_ptr<DataType> ARROW_EXPORT month_interval();
```
nit: as far as naming I could see an argument for using

* `interval_duration`
* `interval_day_time`
* `interval_month`

but I'm OK with the way things are here

emkornfield replied:
I could see this to, but I'm happier without the reverse notation.

emkornfield replied:
note "duration" conflicts with some of our time classes so I renamed to "duration_type", hopefully that is ok

wesm replied:
"some of our time classes"?

emkornfield replied:
Need to look back I think it was in the vendored time library we were using.  I can look deeper if you don't think there should be a conflict

emkornfield replied:
looking again they shouldn't conflict, I will try to fix this today.

wesm replied:
Awesome, thanks.

emkornfield replied:
renamed.  had to touch some of the vendored code.

wesm commented on format/Schema.fbs:
```diff
@@ -183,10 +183,35 @@ table Timestamp {
+//   4-byte integers. 
+// DAY_TIME - Indicates the number of elapsed days and milliseconds,
+//   stored as 2 contiguous 32-bit integers (8-bytes in total).  Support
+//   of this IntervalUnit is not required for full arrow compatibility.
```
Got it. If it isn't being deprecated then it's reasonable to implement it in C++ and have integration tests

emkornfield commented:
@wesm Thanks for the review I'll try to address comments in the next few days.  In terms of naming, I don't know the etiquette for changing the name once it gets approved on the ML.  If was going to change it, I think Duration would be my choice.

@pravindra already reviewed most of the java code, the main changes that happened where plumbing through support for the integration test.

emkornfield commented on java/vector/src/main/codegen/data/ValueVectorTypes.tdd:
```diff
@@ -76,6 +76,15 @@
+        { class: "DurationIntervalMicro", javaType: "long", friendlyType: "Duration",
+          arrowType: "org.apache.arrow.vector.types.pojo.ArrowType.DurationInterval"}
+        { class: "DurationIntervalNano", javaType: "long", friendlyType: "Duration",
+          arrowType: "org.apache.arrow.vector.types.pojo.ArrowType.Duration"}
```
Potentially, I followed the pattern already established in the codebase, but I'm open to trying if it is desirable.  @pravindra @siddharthteotia do you think this is desirable/possible?

jacques-n replied:
Sounds okay to me. Let's see what @siddharthteotia and @pravindra say.

pravindra replied:
IMO, having a single class is definitely desirable. Single this one is starting on a fresh slate, it's worth a try.

kou commented on cpp/src/arrow/array/builder_dict.cc:
```diff
@@ -349,6 +355,9 @@ template class DictionaryBuilder<Date64Type>;
 template class DictionaryBuilder<TimestampType>;
+template class DictionaryBuilder<DurationIntervalType>;
+template class DictionaryBuilder<DayTimeIntervalType>;
+template class DictionaryBuilder<MonthIntervalType>;
```
Just for your information.

These changes will be conflicted with #4255.
If #4255 is merged, we will be able to just remove these changes.

emkornfield replied:
Thanks, they were removed.

emkornfield commented:
@wesm I think I addressed all your feedback.  The OS/X build seems to in an infinite CMake loop with something like 

```For compatibility with older versions of CMake, the install_name fields for
  the following targets are still affected by RPATH settings:
   arrow_python_shared
   arrow_shared
   arrow_testing_shared
   parquet_shared
   plasma_shared
```
I'm not sure why my change would have caused this.  Any thoughts?

@pravindra I was able to consolidate all the java vectors into one, could you take a look?

wesm commented:
@pravindra @emkornfield where do things stand in this patch? Since I have some changes incoming in ipc/json-internal.cc and ipc/metadata-internal.cc it would be good to merge soon so I can address rebase conflicts

pravindra commented:
the changes lgtm.

emkornfield commented:
Thanks @pravindra.  @wesm if you are happy with the changes I think it can merged.

emkornfield commented:
Also if there is a way to force appveyor to run it would make me feel a bit more comfortable as windows has bitten me in the past

wesm commented:
@emkornfield usually a `git commit --amend && git push -f` will do the trick, do you want to try that?

wesm commented:
Baffling

wesm commented:
It must have something to do with this being an older PR. Is Appveyor enabled on your fork? If not I can push this to my work to get an Appveyor build going

emkornfield commented:
No appveyor isn't enabled on my fork.  let me try squashing everything, see if that helps.

emkornfield commented:
Ahh, squash seems to have worked, lets hope things stay green.

wesm commented:
@emkornfield it looks you made some changes to vendored code which I think we should try to avoid doing in general, can we open a JIRA to restore to the vendored version and address the name conflict another way? I don't want this issue to block the patch since I need this to go in and then rebase #4316 on top

wesm commented:
Appveyor appears to be about ~6 hours backed up so I'm going to push a branch on my fork to try to get a quicker thumbs up to merge this

wesm commented:
Here's an Appveyor build starting now on my fork which will take a while to run https://ci.appveyor.com/project/wesm/arrow/builds/24577591

emkornfield commented:
Thanks.  https://issues.apache.org/jira/browse/ARROW-5346 opened to revert the changes.  Right now the only solution seems like reordering includes in some cases or fixing upstream.  I will have to think about this.

wesm commented:
Appveyor build looks OK, I'm going to wait a little while yet to make sure the MinGW build works before merging

emkornfield commented:
@wesm appveyor on your branch looks all greeen.

wesm commented:
Awesome. Thanks @emkornfield!!
